### PR TITLE
feat: support anonymization/pseudonymization for conditional references

### PR DESCRIPTION
**Pull request by jabberwoc**
See Issue [Conditional references #50](https://github.com/miracum/fhir-pseudonymizer/issues/50).

### CODE REVIEW

**Comment by jabberwoc** [comment]
Any comments on this?

**Comment by jabberwoc** [comment]
Now supports identifier search parameters in `IfNoneExist` properties of Bunde entry requests (Conditional creates).

File: src/FhirPseudonymizer/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Utility/ReferenceUtility.cs [hunk 1]
```diff
@@ -9,22 +10,30 @@ public class ReferenceUtility
...
         {
-            // Regex for absolute or relative url reference, https://www.hl7.org/fhir/references.html#literal
+            // Regex for absolute or relative literal url reference, https://www.hl7.org/fhir/references.html#literal
             new Regex(@"^(?<prefix>((http|https)://([A-Za-z0-9\\\/\.\:\%\$])*)?("
                       + string.Join("|", ModelInfo.SupportedResources)
                       + @")\/)(?<id>[A-Za-z0-9\-\.]{1,64})(?<suffix>\/_history\/[A-Za-z0-9\-\.]{1,64})?$"),
+            // Regex for conditional references (https://www.hl7.org/fhir/http.html#trules) or search parameters with identifier
+            new Regex(@"^(?<prefix>(("
+                      + string.Join("|", ModelInfo.SupportedResources)
+                      + @")\?)?identifier=((http|https)://([A-Za-z0-9\\\/\.\:\%\$\-])*\|)?)(?<id>[A-Za-z0-9\-\.]{1,64})$")
```
Comment (chgl) [inline src/FhirPseudonymizer/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Utility/ReferenceUtility.cs:22]: conditional references could also be more complicated by chaining multiple search parameters like so: `ResearchSubject?patient.identifier=http://example.com/patient|123&study.identifier=http://example.com/study|456` which currently doesn't seem to be handled.

Comment (chgl) [inline src/FhirPseudonymizer/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Utility/ReferenceUtility.cs:22]: identifier.system could also be an oid or uuid reference.

**Comment by chgl** [comment]
Thanks for the PR! There's only a conflict in the Dockerfile which you can address by rebasing on master.

Do you think there will be any breaking changes, ie. things that previously weren't pseudonymized now are or things that were pseudonymized now no longer are? (We should probably add some sort of snapshot testing to better protect against such regressions eventually). If no breaking changes are expected I'd be OK with merging despite my comments and simply iterating on the feature later on since the raised concerns may not actually affect the users (you) at this point. Putting this feature behind a feature flag to avoid any unintended changes might also be an option for peace of mind.

Off-topic, but I think it might be better to eventually get rid of the copy-pasted `Microsoft.Health.Fhir.Anonymizer.Shared.Core` code and simply use the upstream nuget package: https://github.com/microsoft/Tools-for-Health-Data-Anonymization/issues/74. The solution implemented here could also be implemented by combining the generalize method which matches on an `ifNoneExist` identifier with a pseudonymize method. There's currently a PR open which allows extending the generalize method: https://github.com/microsoft/Tools-for-Health-Data-Anonymization/pull/169.

File: src/FhirPseudonymizer/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Utility/ReferenceUtility.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jabberwoc) [reply src/FhirPseudonymizer/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Utility/ReferenceUtility.cs:22]: Yes, that was just stupid copy&paste on my part. 😬

Comment (jabberwoc) [reply src/FhirPseudonymizer/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Utility/ReferenceUtility.cs:22]: I didn't think of that, actually. This should be fixed along with the arbitrary Identifier.value length.

**Comment by jabberwoc** [comment]
> Do you think there will be any breaking changes, ie. things that previously weren't pseudonymized now are or things that were pseudonymized now no longer are? (We should probably add some sort of snapshot testing to better protect against such regressions eventually). If no breaking changes are expected I'd be OK with merging despite my comments and simply iterating on the feature later on since the raised concerns may not actually affect the users (you) at this point. Putting this feature behind a feature flag to avoid any unintended changes might also be an option for peace of mind.

I don't think this will be a breaking change. However, you made a valid point that this feature isn't quite ready as a general solution.
I was basically trying to solve a specific problem which is (possibly) the most common use case: Conditional references using identifiers and urls. At least, this is how we build our resources so that was my motivation behind this, obviously.

We use a separate fork of this project and have this feature in production for 5 months now. Along with other changes which are probably even more specific to our needs.
So there is currently no pressure to have this merged quickly from our side.

> Off-topic, but I think it might be better to eventually get rid of the copy-pasted `Microsoft.Health.Fhir.Anonymizer.Shared.Core` code and simply use the upstream nuget package: [microsoft/Tools-for-Health-Data-Anonymization#74](https://github.com/microsoft/Tools-for-Health-Data-Anonymization/issues/74). The solution implemented here could also be implemented by combining the generalize method which matches on an `ifNoneExist` identifier with a pseudonymize method. There's currently a PR open which allows extending the generalize method: [microsoft/Tools-for-Health-Data-Anonymization#169](https://github.com/microsoft/Tools-for-Health-Data-Anonymization/pull/169).

That sounds promising. I will have a look at it when time permits. 😉

**Comment by chgl** [comment]
@jabberwoc I've rebased and put the conditional reference pseudonymization behind a feature flag (arguably excessive given that the change shouldn't really cause anything backwards-incompatible...). You can enable it by setting `Features__ConditionalReferencePseudonymization=true`. You can address the comments above in a seperate PR should the need ever arise.

But this way you don't have to maintain your own fork and can benefit from the latest upstream features (e.g. https://github.com/miracum/fhir-pseudonymizer#usesystemtextjsonfhirserializer). Let me know if the changes look OK to you and I can go ahead and merge it.

**Comment by jabberwoc** [comment]
@chgl Yes, go ahead. Thanks!

**Review by chgl — APPROVED** [review]

**Pull request merged**